File: NEsper/events/property/NestedPropertyGetter.cs
using System;
using System.Collections.Generic;

using net.esper.compat;
using net.esper.events;

namespace net.esper.events.property
{
	/// <summary>
    /// Getter for one or more levels deep nested properties.
    /// </summary>

    public class NestedPropertyGetter : EventPropertyGetter
    {
        private readonly EventPropertyGetter[] getterChain;
        private readonly BeanEventAdapter beanEventAdapter;

        /// <summary> Ctor.</summary>
        /// <param name="getterChain">is the chain of getters to retrieve each nested property
        /// </param>
        /// <param name="beanEventAdapter">is the chache and factory for event bean types and event wrappers
        /// </param>
        public NestedPropertyGetter(IList<EventPropertyGetter> getterChain, BeanEventAdapter beanEventAdapter)
        {
            this.getterChain = CollectionHelper.ToArray( getterChain ) ;
            this.beanEventAdapter = beanEventAdapter;
        }

        /// <summary>
        /// Return the value for the property in the event object specified when the instance was obtained.
        /// Useful for fast access to event properties. Throws a PropertyAccessException if the getter instance
        /// doesn't match the EventType it was obtained from, and to indicate other property access problems.
        /// </summary>
        /// <param name="eventBean">is the event to get the value of a property from</param>
        /// <returns>value of property in event</returns>
        /// <throws>  PropertyAccessException to indicate that property access failed </throws>
        public virtual Object GetValue(EventBean eventBean)
        {
            Object _value = null;

            for (int i = 0; i < getterChain.Length; i++)
            {
            	_value = getterChain[i].GetValue(eventBean);

                if (_value == null)
                {
                    return null;
                }

                if (i < (getterChain.Length - 1))
                {
                    eventBean = beanEventAdapter.AdapterForBean(_value);
                }
            }

            return _value;
        }
    }
}

File: trunk/esper/src/main/java/net/esper/event/property/NestedPropertyGetter.java
package net.esper.event.property;

import net.esper.event.*;
import java.util.List;

/**
 * Getter for one or more levels deep nested properties.
 */
public class NestedPropertyGetter implements EventPropertyGetter
{
    private final EventPropertyGetter[] getterChain;
    private final BeanEventTypeFactory beanEventTypeFactory;

    /**
     * Ctor.
     * @param getterChain is the chain of getters to retrieve each nested property
     * @param beanEventTypeFactory is the chache and factory for event bean types and event wrappers
     */
    public NestedPropertyGetter(List<EventPropertyGetter> getterChain, BeanEventTypeFactory beanEventTypeFactory)
    {
        this.getterChain = getterChain.toArray(new EventPropertyGetter[0]);
        this.beanEventTypeFactory = beanEventTypeFactory;
    }

    public Object get(EventBean eventBean) throws PropertyAccessException
    {
        Object value = null;

        for (int i = 0; i < getterChain.length; i++)
        {
            value = getterChain[i].get(eventBean);

            if (value == null)
            {
                return null;
            }

            if (i < (getterChain.length - 1))
            {
                EventType type = beanEventTypeFactory.createBeanType(value.getClass().getName(), value.getClass());
                eventBean = new BeanEventBean(value, type);
            }
        }
        return value;
    }

    public boolean isExistsProperty(EventBean eventBean)
    {
        int lastElementIndex = getterChain.length - 1;

        // walk the getter chain up to the previous-to-last element, returning its object value.
        // any null values in between mean the property does not exists
        for (int i = 0; i < getterChain.length - 1; i++)
        {
            Object value = getterChain[i].get(eventBean);

            if (value == null)
            {
                return false;
            }
            else
            {
                EventType type = beanEventTypeFactory.createBeanType(value.getClass().getName(), value.getClass());
                eventBean = new BeanEventBean(value, type);
            }
        }
        
        return getterChain[lastElementIndex].isExistsProperty(eventBean);
    }
}
